## Toronto Neighbourhoods Segmentation and Clustering

<img src="https://www.toronto.ca/wp-content/uploads/2017/07/8e31-explore-visitor-banner.jpg" alt="Toronto Neighborhoods" align="left">

<p>&nbsp;</p>
<p><strong>The goal of this Notebook is to practice my Python skills:</strong></p>
<ul>
<li>Using libraries like: pandas, numpy, matplotlib, sklearn, folium & geopy </li>
<li>Processing dataframes & visualizing data using folium </li>
<li>And clustering data using the k-means algorithm. </li>
</ul>

<p><strong>First I used the code to scrape the following Wikipedia page: <a href="https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M ">https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M </a> in order to obtain the data that is in the table of postal codes and to transform the data into a pandas dataframe.</strong></p>

In [0]:
#Install the packages if required (remove #)
#conda install -c conda-forge lxml
#conda install -c anaconda beautifulsoup4

In [1]:
# The data is stored on the Wiki page in structures called wikitables
from pandas.io.html import read_html

page = 'https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'
wikitables = read_html(page, index_col=0, attrs={"class":"wikitable"})

print ("Extracted {num} wikitables".format(num=len(wikitables)))

Extracted 1 wikitables


In [3]:
toronto_postal_codes = wikitables[0]
toronto_postal_codes.head()

Borough     Neighbourhood
Postcode                                    
M1A           Not assigned      Not assigned
M2A           Not assigned      Not assigned
M3A             North York         Parkwoods
M4A             North York  Victoria Village
M5A       Downtown Toronto      Harbourfront

In [4]:
toronto_postal_codes.shape

(288, 2)

<p>&nbsp;</p>
<p><strong>Next I started processing the dataframe according to the Coursera assignment <span style="color: #339966;">instructions</span> mentionen below:</strong></p>
<ul>
<li><span style="color: #339966;">The dataframe will consist of three columns: PostalCode, Borough, and Neighborhood</span></li>
</ul>
<p>&nbsp;</p>

In [5]:
toronto_postal_codes.reset_index(inplace=True)
toronto_postal_codes.columns

Index(['Postcode', 'Borough', 'Neighbourhood'], dtype='object')

<p><span style="color: #339966;">&nbsp;</span></p>
<ul>
<li><span style="color: #339966;">Only process the cells that have an assigned borough. Ignore cells with a borough that is&nbsp;<strong>Not assigned.</strong></span></li>
</ul>
<p>&nbsp;</p>

In [6]:
condition = toronto_postal_codes[ toronto_postal_codes['Borough'] == 'Not assigned' ].index
 # Delete these row indexes from dataFrame
toronto_postal_codes.drop(condition , inplace=True)
toronto_postal_codes.head()

Postcode           Borough     Neighbourhood
2      M3A        North York         Parkwoods
3      M4A        North York  Victoria Village
4      M5A  Downtown Toronto      Harbourfront
5      M5A  Downtown Toronto       Regent Park
6      M6A        North York  Lawrence Heights

<p>&nbsp;</p>
<ul>
<li><span style="color: #339966;">More than one neighborhood can exist in one postal code area. For example, in the table on the Wikipedia page, you will notice that&nbsp;<strong>M5A</strong>&nbsp;is listed twice and has two neighborhoods:&nbsp;<strong>Harbourfront&nbsp;</strong>and&nbsp;<strong>Regent Park</strong>. These two rows will be combined into one row with the neighborhoods separated with a comma as shown in&nbsp;<strong>row 11&nbsp;</strong>in the above table.</span></li>
</ul>
<p>&nbsp;</p>

In [7]:
toronto_postal_codes = toronto_postal_codes.groupby(['Postcode','Borough'])['Neighbourhood'].apply(', '.join).reset_index()
toronto_postal_codes.head()

Postcode      Borough                           Neighbourhood
0      M1B  Scarborough                          Rouge, Malvern
1      M1C  Scarborough  Highland Creek, Rouge Hill, Port Union
2      M1E  Scarborough       Guildwood, Morningside, West Hill
3      M1G  Scarborough                                  Woburn
4      M1H  Scarborough                               Cedarbrae

<p>&nbsp;</p>
<ul>
<li><span style="color: #339966;">If a cell has a borough but a&nbsp;<strong>Not assigned&nbsp;</strong>neighborhood, then the neighborhood will be the same as the borough. So for the&nbsp;<strong>9th</strong>&nbsp;cell in the table on the Wikipedia page, the value of the Borough and the Neighborhood columns will be&nbsp;<strong>Queen's Park.</strong></span></li>
</ul>
<p>&nbsp;</p>

In [8]:
condition = toronto_postal_codes[ toronto_postal_codes['Neighbourhood'] == 'Not assigned' ].index

for i in condition:
    toronto_postal_codes.loc[i]['Neighbourhood'] = toronto_postal_codes.loc[i]['Borough']

toronto_postal_codes.loc[(toronto_postal_codes['Postcode'] == 'M7A')]

Postcode       Borough Neighbourhood
85      M7A  Queen's Park  Queen's Park

<p><span style="color: #339966;">&nbsp;</span></p>
<ul>
<li><span style="color: #339966;">In the last cell of your notebook, use the&nbsp;<strong>.shape</strong>&nbsp;method to print the number of rows of your dataframe.</span></li>
</ul>
<p><span style="color: #339966;">&nbsp;</span></p>

In [9]:
toronto_postal_codes.shape

(103, 3)

<p>&nbsp;</p><p><strong>Next I got the latitude and the longitude coordinates of each neighborhood in order to utilize the Foursquare location data.</strong></p><p>&nbsp;</p>

In [10]:
# I tried to use geocoder with no luck. Will use the CSV file offered by Coursera
import pandas as pd
df_coord = pd.read_csv("https://cocl.us/Geospatial_data")
df_coord.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

In [11]:
toronto_postal_codes = toronto_postal_codes.join(df_coord.set_index('Postal Code'), on='Postcode')
toronto_postal_codes.head()

Postcode      Borough                           Neighbourhood   Latitude  \
0      M1B  Scarborough                          Rouge, Malvern  43.806686   
1      M1C  Scarborough  Highland Creek, Rouge Hill, Port Union  43.784535   
2      M1E  Scarborough       Guildwood, Morningside, West Hill  43.763573   
3      M1G  Scarborough                                  Woburn  43.770992   
4      M1H  Scarborough                               Cedarbrae  43.773136   

   Longitude  
0 -79.194353  
1 -79.160497  
2 -79.188711  
3 -79.216917  
4 -79.239476

<p>&nbsp;</p><p><strong>Finally let's explore and cluster the neighborhoods in Toronto.</strong></p>
<ul>
<li> Geting the latitude and longitude values of Toronto using geopy library.</li>
<li> Creating a map of Toronto with neighborhoods superimposed on top.</li>
</ul>
<p><strong>I am using postcode instead of neighbourhoods since it's easier to read. </strong></p>
<p>&nbsp;</p>

In [17]:
#Install the packages if required (remove #)
#!conda install -c conda-forge geopy --yes 
#!conda install -c conda-forge folium=0.5.0 --yes

In [12]:
# Downloading the dependencies that will be required
import numpy as np  # library to handle data in a vectorized manner
import json # library to handle JSON files
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values
import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe
import matplotlib.cm as cm # Matplotlib and associated plotting modules
import matplotlib.colors as colors
from sklearn.cluster import KMeans # import k-means from clustering stage
import folium # map rendering library

print('Libraries imported.')

Libraries imported.


In [13]:
address = 'Toronto, Canada'

geolocator = Nominatim(user_agent="tor_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.653963, -79.387207.


In [14]:
# Creating a map of Toronto using latitude and longitude values
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

# Adding markers to map
for lat, lng, borough, postcode in zip(toronto_postal_codes['Latitude'], toronto_postal_codes['Longitude'], toronto_postal_codes['Borough'], toronto_postal_codes['Postcode']):
    label = '{}, {}'.format(postcode, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

<p>&nbsp;</p><ul>
<li> Utilizing the Foursquare API to explore the neighborhoods and segment them. </li>
</ul><p>&nbsp;</p>

In [19]:
# My Foursquare credentials were stored locally
config = json.load(open('C:/Users/lcuzacov/Desktop/GitHub/Coursera_Capstone/config.json'))
CLIENT_ID = config['CLIENT_ID']
CLIENT_SECRET = config['CLIENT_SECRET']
VERSION = config['VERSION']

In [21]:
# Creating a function to extract venues information of a neighbourhood in Toronto
radius = 500
LIMIT=100

def getNearbyVenues(names, latitudes, longitudes, radius=500):
   
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        #print(name)
                    
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()['response']['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Postcode', 
                  'Neighbourhood Latitude', 
                  'Neighbourhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [22]:
# Running the code above for all neighbourhoods in Toronto
toronto_venues = getNearbyVenues(names=toronto_postal_codes['Postcode'],
                                   latitudes=toronto_postal_codes['Latitude'],
                                   longitudes=toronto_postal_codes['Longitude']
                                  )

In [23]:
print(toronto_venues.shape)
toronto_venues.head()

(2262, 7)


Postcode  Neighbourhood Latitude  Neighbourhood Longitude  \
0      M1B               43.806686               -79.194353   
1      M1C               43.784535               -79.160497   
2      M1C               43.784535               -79.160497   
3      M1E               43.763573               -79.188711   
4      M1E               43.763573               -79.188711   

                             Venue  Venue Latitude  Venue Longitude  \
0                          Wendy's       43.807448       -79.199056   
1            Royal Canadian Legion       43.782533       -79.163085   
2        Affordable Toronto Movers       43.787919       -79.162977   
3  Swiss Chalet Rotisserie & Grill       43.767697       -79.189914   
4                G & G Electronics       43.765309       -79.191537   

         Venue Category  
0  Fast Food Restaurant  
1                   Bar  
2         Moving Target  
3           Pizza Place  
4     Electronics Store

<p>&nbsp;</p><ul>
Looks like there is a total of 2262 venues that have been extracted using FourSquare.
</ul><p>&nbsp;</p>

In [24]:
toronto_venues.groupby('Postcode').count()

Neighbourhood Latitude  Neighbourhood Longitude  Venue  \
Postcode                                                           
M1B                            1                        1      1   
M1C                            2                        2      2   
M1E                            7                        7      7   
M1G                            4                        4      4   
M1H                            7                        7      7   
M1J                            3                        3      3   
M1K                            5                        5      5   
M1L                           10                       10     10   
M1M                            2                        2      2   
M1N                            4                        4      4   
M1P                            7                        7      7   
M1R                            6                        6      6   
M1S                            4                        4      4   
M1T                           12                       12     12   
M1V                            3                        3      3   
M1W                           13                       13     13   
M2H                            4                        4      4   
M2J                           68                       68     68   
M2K                            4                        4      4   
M2N                           34                       34     34   
M2P                            4                        4      4   
M2R                            6                        6      6   
M3A                            3                        3      3   
M3B                            6                        6      6   
M3C                           22                       22     22   
M3H                           20                       20     20   
M3J                            7                        7      7   
M3K                            2                        2      2   
M3L                            5                        5      5   
M3M                            4                        4      4   
...                          ...                      ...    ...   
M6B                            5                        5      5   
M6C                            3                        3      3   
M6E                            5                        5      5   
M6G                           15                       15     15   
M6H                           15                       15     15   
M6J                           68                       68     68   
M6K                           23                       23     23   
M6L                            4                        4      4   
M6M                            4                        4      4   
M6N                            4                        4      4   
M6P                           25                       25     25   
M6R                           14                       14     14   
M6S                           31                       31     31   
M7A                           39                       39     39   
M7R                           11                       11     11   
M7Y                           16                       16     16   
M8V                           14                       14     14   
M8W                            9                        9      9   
M8X                            2                        2      2   
M8Y                            3                        3      3   
M8Z                           14                       14     14   
M9B                            1                        1      1   
M9C                            7                        7      7   
M9L                            3                        3      3   
M9M                            2                        2      2   
M9N                            2                        2      2  

In [25]:
print('There are {} unique venue categories.'.format(len(toronto_venues['Venue Category'].unique())))

There are 274 unique venue categories.


<p>&nbsp;</p>
<p><strong>Proceeding to analyze the neighbourhoods based on the venue info exported.</strong></p>
<p>&nbsp;</p>

In [26]:
# Using the one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")
toronto_onehot['Postcode'] = toronto_venues['Postcode'] 
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]
toronto_onehot.head()

Postcode  Accessories Store  Afghan Restaurant  Airport  Airport Food Court  \
0      M1B                  0                  0        0                   0   
1      M1C                  0                  0        0                   0   
2      M1C                  0                  0        0                   0   
3      M1E                  0                  0        0                   0   
4      M1E                  0                  0        0                   0   

   Airport Gate  Airport Lounge  Airport Service  Airport Terminal  \
0             0               0                0                 0   
1             0               0                0                 0   
2             0               0                0                 0   
3             0               0                0                 0   
4             0               0                0                 0   

   American Restaurant  ...  Transportation Service  \
0                    0  ...                       0   
1                    0  ...                       0   
2                    0  ...                       0   
3                    0  ...                       0   
4                    0  ...                       0   

   Vegetarian / Vegan Restaurant  Video Game Store  Video Store  \
0                              0                 0            0   
1                              0                 0            0   
2                              0                 0            0   
3                              0                 0            0   
4                              0                 0            0   

   Vietnamese Restaurant  Warehouse Store  Wine Bar  Wings Joint  \
0                      0                0         0            0   
1                      0                0         0            0   
2                      0                0         0            0   
3                      0                0         0            0   
4                      0                0         0            0   

   Women's Store  Yoga Studio  
0              0            0  
1              0            0  
2              0            0  
3              0            0  
4              0            0  

[5 rows x 275 columns]

In [27]:
toronto_onehot.shape

(2262, 275)

In [28]:
# Grouping rows by postalcode and by taking the mean of the frequency of occurrence of each category
toronto_grouped = toronto_onehot.groupby('Postcode').mean().reset_index()
toronto_grouped

0.0   
94      M9N                0.0                0.0      0.0   
95      M9P                0.0                0.0      0.0   
96      M9R                0.0                0.0      0.0   
97      M9V                0.0                0.0      0.0   
98      M9W                0.0                0.0      0.0   

    Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0                  0.0           0.0             0.0              0.0   
1                  0.0           0.0             0.0              0.0   
2                  0.0           0.0             0.0              0.0   
3                  0.0           0.0             0.0              0.0   
4                  0.0           0.0             0.0              0.0   
5                  0.0           0.0             0.0              0.0   
6                  0.0           0.0             0.0              0.0   
7                  0.0           0.0             0.0              0.0   
8                  0.0           0.0             0.0              0.0   
9                  0.0           0.0             0.0              0.0   
10                 0.0           0.0             0.0              0.0   
11                 0.0           0.0             0.0              0.0   
12                 0.0           0.0             0.0              0.0   
13                 0.0           0.0             0.0              0.0   
14                 0.0           0.0             0.0              0.0   
15                 0.0           0.0             0.0              0.0   
16                 0.0           0.0             0.0              0.0   
17                 0.0           0.0             0.0              0.0   
18                 0.0           0.0             0.0              0.0   
19                 0.0           0.0             0.0              0.0   
20                 0.0           0.0             0.0              0.0   
21                 0.0           0.0             0.0              0.0   
22                 0.0           0.0             0.0              0.0   
23                 0.0           0.0             0.0              0.0   
24                 0.0           0.0             0.0              0.0   
25                 0.0           0.0             0.0              0.0   
26                 0.0           0.0             0.0              0.0   
27                 0.0           0.0             0.0              0.0   
28                 0.0           0.0             0.0              0.0   
29                 0.0           0.0             0.0              0.0   
..                 ...           ...             ...              ...   
69                 0.0           0.0             0.0              0.0   
70                 0.0           0.0             0.0              0.0   
71                 0.0           0.0             0.0              0.0   
72                 0.0           0.0             0.0              0.0   
73                 0.0           0.0             0.0              0.0   
74                 0.0           0.0             0.0              0.0   
75                 0.0           0.0             0.0              0.0   
76                 0.0           0.0             0.0              0.0   
77                 0.0           0.0             0.0              0.0   
78                 0.0           0.0             0.0              0.0   
79                 0.0           0.0             0.0              0.0   
80                 0.0           0.0             0.0              0.0   
81                 0.0           0.0             0.0              0.0   
82                 0.0           0.0             0.0              0.0   
83                 0.0           0.0             0.0              0.0   
84                 0.0           0.0             0.0              0.0   
85                 0.0           0.0             0.0              0.0   
86                 0.0           0.0             0.0              0.0   
87                 0.0          

In [29]:
toronto_grouped.shape

(99, 275)

In [35]:
#Creating a function to sort the venues in descending order
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [36]:
# Creating the new dataframe and displaing the top 10 venues for each neighborhood
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Postcode']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Postcode'] = toronto_grouped['Postcode']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Postcode 1st Most Common Venue 2nd Most Common Venue  \
0      M1B  Fast Food Restaurant           Yoga Studio   
1      M1C         Moving Target                   Bar   
2      M1E           Pizza Place        Breakfast Spot   
3      M1G           Coffee Shop     Korean Restaurant   
4      M1H      Hakka Restaurant                Bakery   

         3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0  Eastern European Restaurant        Discount Store               Dog Run   
1                  Yoga Studio    Dim Sum Restaurant        Discount Store   
2            Electronics Store        Medical Center   Rental Car Location   
3           Mexican Restaurant           Yoga Studio               Dog Run   
4                         Bank    Athletics & Sports       Thai Restaurant   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0      Doner Restaurant            Donut Shop             Drugstore   
1               Dog Run      Doner Restaurant            Donut Shop   
2          Intersection    Mexican Restaurant           Yoga Studio   
3      Doner Restaurant            Donut Shop             Drugstore   
4  Caribbean Restaurant   Fried Chicken Joint           Yoga Studio   

  9th Most Common Venue       10th Most Common Venue  
0   Dumpling Restaurant            Electronics Store  
1             Drugstore          Dumpling Restaurant  
2      Doner Restaurant                        Diner  
3   Dumpling Restaurant  Eastern European Restaurant  
4            Donut Shop                      Dog Run

<p>&nbsp;</p>
<p><strong>Clustering the Neighbourhoods</strong></p>
<p>&nbsp;</p>

In [37]:
# set number of clusters
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Postcode', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10]

array([2, 0, 0, 0, 0, 0, 0, 0, 0, 0])

In [38]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

toronto_merged = toronto_postal_codes

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Postcode'), on='Postcode')

toronto_merged.head() # check the last columns!

Postcode      Borough                           Neighbourhood   Latitude  \
0      M1B  Scarborough                          Rouge, Malvern  43.806686   
1      M1C  Scarborough  Highland Creek, Rouge Hill, Port Union  43.784535   
2      M1E  Scarborough       Guildwood, Morningside, West Hill  43.763573   
3      M1G  Scarborough                                  Woburn  43.770992   
4      M1H  Scarborough                               Cedarbrae  43.773136   

   Longitude  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0 -79.194353             2.0  Fast Food Restaurant           Yoga Studio   
1 -79.160497             0.0         Moving Target                   Bar   
2 -79.188711             0.0           Pizza Place        Breakfast Spot   
3 -79.216917             0.0           Coffee Shop     Korean Restaurant   
4 -79.239476             0.0      Hakka Restaurant                Bakery   

         3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0  Eastern European Restaurant        Discount Store               Dog Run   
1                  Yoga Studio    Dim Sum Restaurant        Discount Store   
2            Electronics Store        Medical Center   Rental Car Location   
3           Mexican Restaurant           Yoga Studio               Dog Run   
4                         Bank    Athletics & Sports       Thai Restaurant   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0      Doner Restaurant            Donut Shop             Drugstore   
1               Dog Run      Doner Restaurant            Donut Shop   
2          Intersection    Mexican Restaurant           Yoga Studio   
3      Doner Restaurant            Donut Shop             Drugstore   
4  Caribbean Restaurant   Fried Chicken Joint           Yoga Studio   

  9th Most Common Venue       10th Most Common Venue  
0   Dumpling Restaurant            Electronics Store  
1             Drugstore          Dumpling Restaurant  
2      Doner Restaurant                        Diner  
3   Dumpling Restaurant  Eastern European Restaurant  
4            Donut Shop                      Dog Run

In [39]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Postcode'], toronto_merged['Cluster Labels']):
    if np.isnan(cluster) == False:
        label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
        folium.CircleMarker(
            [lat, lon],
            radius=5,
            popup=label,
            color=rainbow[int(cluster)-1],
            fill=True,
            fill_color=rainbow[int(cluster)-1],
            fill_opacity=0.7).add_to(map_clusters)

map_clusters

<p>&nbsp;</p>
<p><strong>Examining Clusters</strong></p>
<p>&nbsp;</p>

In [30]:
#toronto_merged.loc[toronto_merged['Cluster Labels'] == 0, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

<p>&nbsp;</p>
<p><strong>Cluster description here...</strong></p>
<p>&nbsp;</p>

In [31]:
#toronto_merged.loc[toronto_merged['Cluster Labels'] == 1, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

<p>&nbsp;</p>
<p><strong>Cluster description here...</strong></p>
<p>&nbsp;</p>

In [32]:
#toronto_merged.loc[toronto_merged['Cluster Labels'] == 2, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

<p>&nbsp;</p>
<p><strong>Cluster description here...</strong></p>
<p>&nbsp;</p>

In [33]:
#toronto_merged.loc[toronto_merged['Cluster Labels'] == 3, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

<p>&nbsp;</p>
<p><strong>Cluster description here...</strong></p>
<p>&nbsp;</p>

In [34]:
#toronto_merged.loc[toronto_merged['Cluster Labels'] == 4, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

<p>&nbsp;</p>
<p><strong>Cluster description here...</strong></p>
<p>&nbsp;</p>